### ISSUE TITLE
Generate printout for EU digital certificates

### ISSUE THREAD

**Issue by Balze2** [corona-warn-app/cwa-wishlist#581]
## Description

Generate a printout for the stored EU COVID certificates.

## Suggested Enhancement

*  PDF can be generated for all types of certificates
*  Original EU template used for printing
*  PDF can be shared using the standard OS sharing functionality

----
Currently it's only possible to store a EU digital COVID-Testcertificate in the App.
It would be a great benefit to print a certificate. 

Reason to ask for it:
My daugther is going for a scandinavia trip with a group of 18 persons. (14 children, 4 adults)
The adults have to take care, that the certificates of each child is available.
What if one of the children loose/broke/... its smartphone?
It would be a great backup to have a printed version available.

Thanks for taking into account

----
Internal Tracking-ID: [EXPOSUREAPP-8352](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8352) obsolete (vaccination DCC)
Internal Tracking ID: [EXPOSUREAPP-8328](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8328) obsolete (test DCC)
Internal Tracking ID: [EXPOSUREAPP-8384](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8384) export function for all DCCs - iOS
Internal Tracking ID: [EXPOSUREAPP-8900](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8900) export function for all DCCs - Android

Currently the following is being considered:
PDF can be generated for all types of DCCs with different templates (vaccination, test, recovery)
PDF can be generated only for DCCs issued in Germany
for DCCs issued by another country, an error message is presented to the user 
Original EU template used for printing 
PDF can be shared using the standard OS sharing functionality

related issues and PR:
 corona-warn-app/cwa-quick-test-backend#143 
 corona-warn-app/cwa-server#1491

**Comment by MikeMcC399** [comment]
@Balze2 
I like your idea!

- In the meantime, as a workaround you could take a screenshot from the Corona-Warn-App and print it. 
- Certificates which are issued on paper could be copied. 
- Ones which are issued in PDF format can be printed out as a backup.

**Comment by Balze2** [comment]
@MikeMcC399 
Thank you for your positive reply.

- The workaround using a printed screenshot is no problem for me. But not all parents are able to handle the workflow. :)
- Sure, printed certificates could be coppied, but vacciend people do get a printed document (if they want to).
- The PDF could be copied too. The PDF I received does not look like the vacciened-certificate.
Best regards
Balze2

**Comment by heinezen** [comment]
@Balze2 

Thanks for your proposal. I have informaed the dev team about it. The developers and RKI will think about your proposal. We will report back here on Github if a decision has been made.

----
Corona-Warn-App Open Source Team

**Comment by Balze2** [comment]
Thank you for taking it into account!

**Comment by dsarkar** [comment]
Internal Tracking-ID: [EXPOSUREAPP-8352](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8352) obsolete (vaccination DCC)
Internal Tracking ID: [EXPOSUREAPP-8328](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8328) obsolete (test DCC)
Internal Tracking ID: [EXPOSUREAPP-8384](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8384) export function for all DCCs


Currently the following is being considered:
- PDF can be generated for all types of certificates
- Original EU template used for printing
- PDF can be shared using the standard OS sharing functionality

**Comment by Ein-Tim** [comment]
I assume we will see something like this in one of the next releases: https://github.com/corona-warn-app/cwa-quick-test-backend/pull/143 & https://github.com/corona-warn-app/cwa-server/pull/1491

**Comment by Ein-Tim** [comment]
iOS PR: https://github.com/corona-warn-app/cwa-app-ios/pull/3386

**Comment by ndegendogo** [comment]
@Ein-Tim do you understand the description of the PR? What does he mean with "show a printable version"?

**Comment by ndegendogo** [comment]
of course a mobile device cannot print.
But instead of *showing* a printable version I would expect to *export* a printable version?

**Comment by Ein-Tim** [comment]
@ndegendogo 

The printable version is the QR-Code which is shown in the app on the corresponding template from here: https://github.com/Digitaler-Impfnachweis/certification-apis/blob/master/templates/README.md

**Comment by ndegendogo** [comment]
Ah - thanks. But how do I then print it?

**Comment by Ein-Tim** [comment]
@ndegendogo 

> of course a mobile device cannot print.
But instead of showing a printable version I would expect to export a printable version?

In German the text in the app is "Druckversion anzeigen", similar to what you can do when creating an Event QR code.

**Comment by Ein-Tim** [comment]
> Ah - thanks. But how do I then print it?

I assume like you can print the QR code from an event at the moment.

**Comment by ndegendogo** [comment]
So maybe they think "exportieren" is too technical, and the concept of showing it is easier to understand for non-technical people? 😅

**Comment by jucktnich** [comment]
You can print from an iPhone indeed. Just search for "AirPrint"

**Comment by Ein-Tim** [comment]
One recommendation: Don't call it "Druckversion" but rather "Zertifikat Exportieren".

---

iOS PRs reg. this have the milestone [2.10](https://github.com/corona-warn-app/cwa-app-ios/milestone/27?closed=1) attached to it, this does not seem to match with the `Fix 2.9` label here.

**Comment by ndegendogo** [comment]
> One recommendation: Don't call it "Druckversion" but rather "Zertifikat Exportieren".

😅😅😅

@Ein-Tim Maybe I can display the "Druckversion", then make a screenshot, and export that to my laptop for printing

**Comment by Ein-Tim** [comment]
@ndegendogo If I understood correctly you will see a preview of the "Druckversion" and then you can share it with the standard iOS / Android sharing menu.

**Comment by ndegendogo** [comment]
@Ein-Tim sounds good 🎉

**Comment by idoodler** [comment]
I would appreciate this feature, we are about to depart from Iceland and Iceland Air requires Covid-19 documents, but they don't accept Screenshots of the QR Code, rather then Official Vaccination documents.

So we can't use the online check in, but have to use the personal checkin at the airport

**Comment by Ein-Tim** [comment]
@idoodler 

CWA 2.10 will come with this feature. In the meantime you can use [CovPass](https://digitaler-impfnachweis-app.de/en), there you can create print outs of your EU digital COVID certificates already now.

**Comment by Ein-Tim** [comment]
Hey @Balze2!

This feature has been implemented in version 2.10, which has been released today.

More information can be found in this blog post: 
[Version 2.10: Corona-Warn-App ready to inform users about booster vaccinations](https://www.coronawarn.app/en/blog/2021-09-22-cwa-version-2-10/)

If you consider your request as implemented, I'd like to ask you to close this issue. 

Thank you!

**Comment by alanrick** [comment]
I can confirm that it works as described on my iPhone. But perhaps someone has the patience to explain to me why this is the best possible implementation. 

I’m confused by the text on the pdf generated that states I can import this pdf/paper version into my CWA to obtain a digital version. Why would I want to export a pdf from CWA and then import it into the CWA again? 

Surely the qr-code rendering of the cert’s character-string alone is the certificate? The fold-instructions/yellow-graphics etc are simply embellishments irrelevant in determining the legitimacy of the test/Immunity. 

My own personal experience was that the different countries I visited that required the EU certificate were satisfied when I uploaded the png rendering of the qr code. And whenever I’ve needed a paper backup a print-out of the screenshot from my Apple Wallet has sufficed. 

Bottom line: Is the qr-code alone sufficient, or are the other text graphic embellishments necessary for validation, too? If the embellishments don’t play a role in the legitimacy, wouldn’t it make more sense for the CWA exported version to explain how to validate the paper print-out (e.g. CovPass Check App) rather than how to reimport it?

ps: many thanks to @Ein-Tim for replying to my twitter comments

**Comment by Balze2** [comment]
Thank you very much for adding this function to CWA!

Best regards

balze2

**Comment by dsarkar** [comment]
Dear @Balze2 and community,

Thanks for contributing to this project. Best wishes, DS 

---
Corona-Warn-App Open Source Team

**Comment by Ein-Tim** [comment]
@alanrick 

> I’m confused by the text on the pdf generated that states I can import this pdf/paper version into my CWA to obtain a digital version. Why would I want to export a pdf from CWA and then import it into the CWA again?

> Surely the qr-code rendering of the cert’s character-string alone is the certificate? The fold-instructions/yellow-graphics etc are simply embellishments irrelevant in determining the legitimacy of the test/Immunity.

These things, the text staying to scan the QR code with CovPass or CWA, as well as the fold instructions are only there because the original template is used. What you create in the CWA is the same as you would get when going to a pharmacy with your yellow vaccination booklet and asking them for a certificate.

**Comment by alanrick** [comment]
But what’s the point in using the same template as the pharmacy? The pharmacy version was generated for people to import the certificate into an app to avoid carrying the yellow booklet with them, but the intention of the generate button is to create a paper backup that can be used when your phones dead/broken. The former  is designed for  importing, but the latter is intended as a „passport“ to be validated.  Different use-case -> different text.

**Comment by MikeMcC399** [comment]
@alanrick 
CWA 2.10.1 generates the standard format EU Digital COVID Certificate for Germany.  This is not a special "pharmacy version". It is the one used by all generated papier format versions.

Users can do what they want with this, for example:
- show it to somebody
- print it out
- archive it digitally
- scan the QR into another device

It does also include the text "Mehr informationen unter: www.digitaler-impfnachweis-app.de" which contains lots of information in German and English about certificates and their usage.

**Comment by alanrick** [comment]
I agree you can use the pdf for all this and more. But the problem to solve in this wish-list item was to generate a paper copy to be used as a “green pass” backup where legitimization by non-digital experts is important. I also realise that my gripe would best be solved by improving the German template to explain how the pass can be validated, because without validating this makes a mockery of a beautifully devised EU green-pass system. 

So far my passes (paper/app/wallet) have been inspected about 40 times but not even the German Border Control validated it electronically. So an extra educational line of text above/below/within the graphic generated by the template explaining how to validate it would have been cool.

As an aside, the only places that _did_ validate my pass were pubs, hotels and restaurants in Dublin. Digital capital of Europe?

**Comment by MikeMcC399** [comment]
@alanrick 
I understand your point about wanting to improve the text on the templates, however this is probably the wrong place to make suggestions to get it changed.

I'm not completely clear about who controls the template formats which are based on EU guidelines. I believe though that Corona-Warn-App is more or less just a consumer of the template.

If you look in the https://github.com/Digitaler-Impfnachweis/certification-apis/tree/master/templates you can find [DCC Print templates (Germany)](https://github.com/Digitaler-Impfnachweis/certification-apis/tree/master/templates). The CWA community made some formatting suggestions for the template in [certification-apis/issues](https://github.com/Digitaler-Impfnachweis/certification-apis/issues) which were picked up and implemented. That repository also hosts [discussions](https://github.com/Digitaler-Impfnachweis/certification-apis/discussions).
